File: python/leetcode/_I_1614_M.py
from typing import List, Dict


class Solution:
    def bestLine(self, points: List[List[int]]) -> List[int]:
        best = [0, 1]
        maxPointNum = 2
        # Dict[float, [firstPointIndex, secondPointIndex, count]]
        lines: Dict[float, List[int]] = {}
        firstI = 0
        pointNum = len(points)
        while firstI < pointNum - maxPointNum:
            lines.clear()
            sameToFirst = 1
            firstP = points[firstI]
            localMaxNum = 0
            localMaxK = None
            for secondI in range(firstI + 1, pointNum):
                secondP = points[secondI]
                xd = secondP[0] - firstP[0]
                yd = secondP[1] - firstP[1]
                if xd == 0:
                    if yd == 0:
                        sameToFirst += 1
                        continue
                    k = float('inf')
                elif yd == 0:
                    k = 0
                else:
                    k = yd / xd
                line = lines.get(k)
                if line is None:
                    lines[k] = line = [firstI, secondI, 0]
                line[2] += 1
                if localMaxNum < line[2]:
                    localMaxNum = line[2]
                    localMaxK = k
            if maxPointNum < sameToFirst + localMaxNum:
                maxPointNum = sameToFirst + localMaxNum
                if localMaxK is None:
                    best = [firstI, firstI + 1]
                else:
                    best = lines[localMaxK][:2]
            firstI += 1
        return best


if __name__ == '__main__':
    print(Solution().bestLine([[0,0],[1,1],[1,0],[2,0]]))

File: javascript/src/leetcode/_I_1614_M.ts

function bestLine(points: [number, number][]) {
    const best: [number, number] = [0, 1];
    let maxPointNum = 2;
    // Map<斜率, [firstPointIndex, secondPointIndex, pointCount]>
    const lines = new Map<number, [number, number, number]>();
    for (let i = 0, len = points.length; i < len - maxPointNum; i++) {
        lines.clear();
        const firstPoint = points[i];
        let sameToFirst = 1;
        let localMaxNum = 0;
        let localMaxK = NaN;
        for (let j = 0; j < len; j++) {
            const secondPoint = points[j];
            const xd = secondPoint[0] - firstPoint[0];
            const yd = secondPoint[1] - firstPoint[1];
            let k: number;
            if (xd == 0) {
                if (yd == 0) {
                    sameToFirst++;
                    continue;
                }
                else {
                    k = Infinity;
                }
            }
            else if (yd == 0) {
                k = 0;
            }
            else {
                k = yd / xd;
            }
            let line = lines.get(k);
            if (line == null) {
                lines.set(k, line = [i, j, 0]);
            }
            line[2]++;
            if (localMaxNum < line[2]) {
                localMaxNum = line[2];
                localMaxK = k;
            }
        }
        if (maxPointNum < sameToFirst + localMaxNum) {
            maxPointNum = sameToFirst + localMaxNum;
            if (isNaN(localMaxK)) {
                best[0] = i;
                best[1] = i + 1;
            }
            else {
                const line = lines.get(localMaxK);
                best[0] = line[0];
                best[1] = line[1];
            }
        }
    }
    return best;
}

console.log(bestLine([[0, 0], [0,0], [0,0]]));
console.log(bestLine([[0,0],[1,1],[1,0],[2,0]]));
